src/main/java/emu/grasscutter/game/combine/CombineManger.java

```diff
@@ -87,11 +92,121 @@ public CombineResult combineItem(Player player, int cid, int count) {
+        result.setBack(new ArrayList<ItemParamData>());
+        result.setRandom(new ArrayList<ItemParamData>());
+
+        // add lucky items
+        if (luckyCount > 0) {
+            switch (combineBonusData.getBonusType()) {
+                case "COMBINE_BONUS_DOUBLE" -> {
+                    var combineExtra = new ItemParamData(combineData.getResultItemId(), luckyCount);
+                    player.getInventory().addItem(combineExtra);
+                    result.getExtra().add(combineExtra);
+                }
+                case "COMBINE_BONUS_REFUND" -> {
+                    if (combineData.getMaterialItems().size() == 1) {
+                        var combineBack =
+                                new ItemParamData(combineData.getMaterialItems().get(0).getItemId(), luckyCount);
+                        player.getInventory().addItem(combineBack);
+                        result.getBack().add(combineBack);
+                    } else {
+                        HashMap<Integer, Integer> mapIdCount = new HashMap<>();
+                        for (int i = 0; i < luckyCount; i++) {
+                            var randomId =
+                                    combineData
+                                            .getMaterialItems()
+                                            .get(
+                                                    ThreadLocalRandom.current()
+                                                            .nextInt(combineData.getMaterialItems().size()))
+                                            .getItemId();
+                            mapIdCount.put(randomId, mapIdCount.getOrDefault(randomId, 0) + 1);
+                        }
+
+                        for (var entry : mapIdCount.entrySet()) {
+                            var combineBack = new ItemParamData(entry.getKey(), entry.getValue());
+                            player.getInventory().addItem(combineBack);
+                            result.getBack().add(combineBack);
+                        }
+                    }
+                }
+                case "COMBINE_BONUS_REFUND_RANDOM" -> {
+                    // for yae miko, "Has a 25% chance to get 1 regional Character Talent Material (base
+                    // material excluded) when crafting. The rarity is that of the base material." from wiki
+                    // map of material id to region id
+                    Map<Integer, Integer> itemToRegion = Map.of(
+                            104301, 1,
+                            104304, 1,
+                            104307, 1,
+                            104310, 2,
+                            104313, 2,
+                            104316, 2,
+                            104320, 3,
+                            104323, 3,
+                            104326, 3,
+                            104329, 4,
+                            104332, 4,
+                            104335, 4
+                    );
```

Comment:
still think this should be moved out of here as a constant somewhere else, or a copyable list if modified here

Reply:
this is also way too many items for `Map.of` to handle, consider making the map with `new HashMap<>() {{}}`